modules/ipp/scanner.go

```diff
@@ -289,45 +297,39 @@ func (scanner *Scanner) augmentWithCUPSData(scan *scan, target *zgrab2.ScanTarge
 	}
 	// Store data into BodyText and BodySHA256 of cupsResp
 	storeBody(cupsResp, scanner)
-	if versionNotSupported(&scan.results.CUPSResponse.BodyText, scanner) {
-		// TODO: Make this step down a version number
+	if versionNotSupported(scan.results.CUPSResponse.BodyText, scan) {
 		return zgrab2.NewScanError(zgrab2.SCAN_APPLICATION_ERROR, ErrVersionNotSupported)
 	}
 
 	bodyBytes := []byte(cupsResp.BodyText)
-	cupsVersions, _ := readAttributeFromBody(CupsVersion, &bodyBytes)
-	if len(cupsVersions) > 0 {
-		// TODO: Include an additional field in ScanResults for each attribute
-		// Report this cupsVersion b/c it's more detailed than the one found in Server header
-		scan.results.CUPSVersion = strings.Split(scan.results.CUPSVersion, "/")[0] + "/" + string(cupsVersions[0])
+	// Write reported CUPS version to results object
+	if cupsVersions, err := readAttributeFromBody(CupsVersion, &bodyBytes); err != nil {
+		scan.log.WithFields(log.Fields{
+			"error": err,
+			"attribute": CupsVersion,
+		}).Warn("Failed to read attribute.")
+	} else if len(cupsVersions) > 0 {
+		scan.results.AttributeCUPSVersion = string(cupsVersions[0])
 	}
-	// TODO: Determine whether ipp-versions must be in increasing order, if not, sort them
-	// TODO: Overwrite ipp_version with highest version listed
-	ippVersions, _ := readAttributeFromBody(VersionsSupported, &bodyBytes)
-	if len(ippVersions) > 0 {
-		highestVersion := string(ippVersions[len(ippVersions)-1])
-		components := strings.Split(highestVersion, ".")
-
-		output, err := strconv.Atoi(components[0])
-		if err != nil {
-			//handle
-		}
-		major := int8(output)
-		output, err = strconv.Atoi(components[1])
-		if err != nil {
-			//handle
-		}
-		minor := int8(output)
-		// Only compare new values to previous if we have previous values
-		comparable := scan.results.MajorVersion != nil && scan.results.MinorVersion != nil
-		if !comparable || (comparable && major >= *scan.results.MajorVersion && minor > *scan.results.MinorVersion) {
-			scan.results.MajorVersion = &major
-			scan.results.MinorVersion = &minor
+	// Write reported IPP versions to results object
+	if ippVersions, err := readAttributeFromBody(VersionsSupported, &bodyBytes); err != nil {
+		scan.log.WithFields(log.Fields{
+			"error": err,
+			"attribute": VersionsSupported,
+		}).Warn("Failed to read attribute.")
+	} else {
+		for _, v := range ippVersions {
+			scan.results.AttributeIPPVersions = append(scan.results.AttributeIPPVersions, string(v))
 		}
 	}
-	uris, _ := readAttributeFromBody(PrinterURISupported, &bodyBytes)
-	if len(uris) > 0 {
-		scan.results.PrinterURI = string(uris[0])
+	// Write reported printer URI to results object
+	if uris, err := readAttributeFromBody(PrinterURISupported, &bodyBytes); err != nil {
+		scan.log.WithFields(log.Fields{
+			"error": err,
+			"attribute": PrinterURISupported,
+		}).Warn("Failed to read attribute.")
+	} else if len(uris) > 0 {
+		scan.results.AttributePrinterURI = string(uris[0])
```

Comment:
Future -- is it permissible for multiple URIs to be returned? If so, it may be helpful to return them all (on the other hand, if it's just an artifact of the API always returning an array, ignore this)